File: php/v2.7/examples/UserRoles/CreateUserRole.php
<?php
/*
 * Copyright 2015 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Require the base class.
require_once dirname(__DIR__) . '/BaseExample.php';

/**
 * This example creates a user role in a given DoubleClick Campaign Manager
 * subaccount. To get the subaccount ID, run GetSubaccounts. To get the
 * available permissions, run GetUserRolePermissions. To get the parent user
 * role ID, run GetUserRoles.
 */
class CreateUserRole extends BaseExample {
  /**
   * (non-PHPdoc)
   * @see BaseExample::getInputParameters()
   * @return array
   */
  protected function getInputParameters() {
    return [['name' => 'user_profile_id',
             'display' => 'User Profile ID',
             'required' => true],
            ['name' => 'parent_user_role_id',
             'display' => 'Parent User Role ID',
             'required' => true],
            ['name' => 'subaccount_id',
             'display' => 'Subaccount ID',
             'required' => true],
            ['name' => 'permission_one',
             'display' => 'First Permission ID',
             'required' => true],
            ['name' => 'permission_two',
             'display' => 'Second Permission ID',
             'required' => true],
            ['name' => 'user_role_name',
             'display' => 'User Role Name',
             'required' => true]];
  }

  /**
   * (non-PHPdoc)
   * @see BaseExample::run()
   */
  public function run() {
    $values = $this->formValues;

    printf(
        '<h2>Creating user role with name "%s" under parent role ID %s</h2>',
        $values['user_role_name'], $values['parent_user_role_id']
    );

    $userRole = new Google_Service_Dfareporting_UserRole();
    $userRole->setName($values['user_role_name']);
    $userRole->setParentUserRoleId($values['parent_user_role_id']);
    $userRole->setPermissions(
        [$values['permission_one'], $values['permission_two']]);
    $userRole->setSubaccountId($values['subaccount_id']);

    $result = $this->service->userRoles->insert(
        $values['user_profile_id'], $userRole
    );

    $this->printResultsTable('User role created.', [$result]);
  }

  /**
   * (non-PHPdoc)
   * @see BaseExample::getName()
   * @return string
   */
  public function getName() {
    return 'Create User Role';
  }

  /**
   * (non-PHPdoc)
   * @see BaseExample::getResultsTableHeaders()
   * @return array
   */
  public function getResultsTableHeaders() {
    return ['id' => 'User Role ID',
            'name' => 'User Role Name'];
  }
}

File: ruby/v2.7/user_roles/create_user_role.rb
#!/usr/bin/env ruby
# Encoding: utf-8
#
# Copyright:: Copyright 2016, Google Inc. All Rights Reserved.
#
# License:: Licensed under the Apache License, Version 2.0 (the "License");
#           you may not use this file except in compliance with the License.
#           You may obtain a copy of the License at
#
#           http://www.apache.org/licenses/LICENSE-2.0
#
#           Unless required by applicable law or agreed to in writing, software
#           distributed under the License is distributed on an "AS IS" BASIS,
#           WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
#           implied.
#           See the License for the specific language governing permissions and
#           limitations under the License.
#
# This example creates a user role in a given DCM account.
#
# To get the account ID, run get_all_userprofiles.rb. To get the parent user
# role ID, run get_user_roles.rb.

require_relative '../dfareporting_utils'
require 'securerandom'

def create_user_role(profile_id, account_id, parent_role_id)
  # Authenticate and initialize API service.
  service = DfareportingUtils.get_service()

  # Create a new user role resource to insert.
  user_role = DfareportingUtils::API_NAMESPACE::UserRole.new({
    :account_id => account_id,
    :name => 'Example User Role #%s' % SecureRandom.hex(3),
    :parent_user_role_id => parent_role_id
  })

  # Insert the user role.
  result = service.insert_user_role(profile_id, user_role)

  # Display results.
  puts 'Created user role with ID %d and name "%s".' % [result.id, result.name]
end

if __FILE__ == $0
  # Retrieve command line arguments.
  args = DfareportingUtils.get_arguments(ARGV, :profile_id, :account_id,
      :parent_role_id)

  create_user_role(args[:profile_id], args[:account_id], args[:parent_role_id])
end
